### ISSUE TITLE
1.3.6rc2: mod_sftp fxp.c segfault, uninitialized stat struct

### ISSUE THREAD

**Issue by qqueue** [proftpd/proftpd#306]
### What I Did

(This bug report is a little weird since I traced it in gdb and "fixed" it a week or two ago, thus I'm recalling details from IRC logs).

I got a segfault with a backtrace of:

```
fxp_get_path_listing (p=0x8baec0, path=0x7dac80 <vwd> "/", st=0x7fffffffde30, user_owner=0x0, group_owner=0x0) at fxp.c:2515
0x00000000005477b8 in fxp_name_write (p=0x8baec0, buf=0x7fffffffdde0, buflen=0x7fffffffddbc, path=0x7dac80 <vwd> "/", st=0x7fffffffde30, user_owner=0x0, group_owner=0x0) at fxp.c:2616
0x000000000055e958 in fxp_handle_realpath (fxp=0x8baf08) at fxp.c:9932
0x0000000000567316 in sftp_fxp_handle_packet (p=0x877ca0, ssh2=0x877ce8, channel_id=0, data=0x877e00 "", datalen=14) at fxp.c:12475
0x000000000053c391 in process_channel_data (chan=0x8bbb58, pkt=0x877ce8, data=0x877e00 "", datalen=14) at channel.c:487
0x000000000053c72e in handle_channel_data (pkt=0x877ce8, channel_id=0x7fffffffe018) at channel.c:572
0x000000000053e4cb in sftp_channel_handle (pkt=0x877ce8, mesg_type=94 '^') at channel.c:1353
```

after much futzing around, it turns out the st struct passed into there was 

```
$20 = {st_dev = 16, st_ino = 256, st_nlink = 9154312, st_mode = 293088000, st_uid = 0, st_gid = 4294958720, __pad0 = 32767, st_rdev = 7702976502244584192, 
st_size = 140737488346784, st_blksize = 4346279, st_blocks = 9154360, st_atim = {tv_sec = 293088000, tv_nsec = 140737488346832}, st_mtim = {tv_sec = 7702976502244584192, 
  tv_nsec = 140737488346832}, st_ctim = {tv_sec = 7702976502244584192, tv_nsec = 140737488346832}, __glibc_reserved = {9154240, 21474836485, 9181833}
```

i.e. random values.

I finally changed this as a hope to debug:

```diff
diff --git a/contrib/mod_sftp/fxp.c b/contrib/mod_sftp/fxp.c
index fba0fbf..067eb43 100644
--- a/contrib/mod_sftp/fxp.c
+++ b/contrib/mod_sftp/fxp.c
@@ -9608,6 +9608,8 @@ static int fxp_handle_realpath(struct fxp_packet *fxp) {
   struct fxp_packet *resp;
   cmd_rec *cmd;
 
+  memset(&st, 0, sizeof(st));
+
   path = sftp_msg_read_string(fxp->pool, &fxp->payload, &fxp->payload_sz);
   if (fxp_session->client_version >= fxp_utf8_protocol_version) {
     path = sftp_utf8_decode_str(fxp->pool, path);
```

i.e. unconditionally zero the st struct, which strangely enough stopped the segfaults, and I got disinterested after that point. My inexpert analysis says that zeroing the struct makes code after that fact correctly detect that it is uninitialized and then will fill in the correct values, rather than leaving them in their whatever-memory-happened-to-be-there-before state.

I'm again hoping that the bug as presented here is straightforward enough to somebody with better C knowledge, but I can try to put together a more minimal test case if needed.

### ProFTPD Version and Configuration
 
Identical to #305 .

**Comment by Castaglia** [comment]
Thanks for the report!

What I find interesting, looking at that particular section of code closely, is that that `struct stat` structure _should_ be being properly populated by the calls to either `pr_fsio_lstat()` or `pr_fsio_stat()`.  If those functions _fail_, then that `fxp_name_write()` call (which doesn't handle the uninitialized data well) isn't called.  And thus if that `stat` call _succeeds_, then I would _assume_ that that `struct stat` now has all-valid data.  But this case shows that that is not the case.

Now, using modules like `mod_vroot` or `mod_statcache` can affect the behavior of those `stat()` functions -- but your `proftpd` build doesn't use those modules.

What SFTP client did you saw you were using, to see this behavior?  I am now suspecting that the culprit here might be the initial value of zero for the `realpath_flags` variable.  There is a case where, for older protocol versions, that initial value of zero is *not* changed.  But the `switch` statement for `realpath_flags` doesn't have any cases for a value of zero -- and those cases are where the `stat` functions are called.  Thus perhaps another possible fix is:

    unsigned char *buf, realpath_flags = SSH2_FXRP_NO_CHECK;

at the beginning of `fxp_handle_realpath`?

**Comment by qqueue** [comment]
I'm connecting with filezilla:

```
FileZilla Client
----------------

Version:          3.15.0.2

Build information:
  Compiled for:   x86_64-pc-linux-gnu
  Compiled on:    x86_64-pc-linux-gnu
  Build date:     2016-02-25
  Compiled with:  gcc (Ubuntu 5.3.1-10ubuntu1) 5.3.1 20160224
  Compiler flags: -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wall -g -std=gnu++14

Linked against:
  wxWidgets:      3.0.2
  GnuTLS:         3.4.10
  SQLite:         3.11.0

Operating system:
  Name:           Linux 4.4.0-31-generic x86_64
  Version:        4.4
  CPU features:   sse sse2 sse3 ssse3 sse4.1 sse4.2 avx avx2 aes pclmulqdq rdrnd bmi2 bmi2
  Settings dir:   /home/queue/.config/filezilla/
```

It turns out this bug is pretty trivially reproducible from the 1.3.6rc2 tarball, compiled with
`./configure --disable-strip --enable-devel=stacktrace --with-modules=mod_tls:mod_tls_shmcache:mod_sftp:mod_sql:mod_shaper:mod_sql_mysql:mod_sftp_sql --enable-nls --enable-ctrls` (probably not minimal, but sufficient)

and this config:

```
ServerName                      "SFTP server"
ServerAdmin                     root@localhost
ServerIdent                     off
DefaultServer                   on

DefaultRoot                     ~
UseReverseDNS                   off

<VirtualHost 127.0.0.1>
SFTPEngine on
TLSEngine off
Port 25725
SFTPOptions InsecureHostKeyPerms
SFTPHostKey /etc/ssh/ssh_host_ecdsa_key
SFTPHostKey /etc/ssh/ssh_host_rsa_key
</VirtualHost>
```

Simply logging in with any system user will trigger the segfault.

Other good news is I just tried initializing `realpath_flags = SSH2_FXRP_NO_CHECK;` at the beginning of `fxp_handle_realpath`, and that does indeed fix the segfault.

**Comment by Castaglia** [comment]
Excellent!  Thanks for the verification.  I'll get that change pushed to master *and* to the 1.3.5 branch right away.

**Comment by Castaglia** [comment]
Fixed in master and 1.3.5.
